katacomb/katacomb/aips_export.py

```diff
@@ -271,34 +271,37 @@ def export_calibration_solutions(uv_files, kat_adapter, mfimage_params, telstate
     for si, uv_file in enumerate(uv_files):
         try:
             with uv_factory(aips_path=uv_file, mode='r') as uvf:
-                try:
-                    log.info('Extracting phase only self-calibration from AIPS SN: %d' % (pSN))
-                    uvf.attach_table("AIPS SN", pSN)
+                log.info('Extracting self-calibration solutions from %s' % uv_file)
+                sn_versions = [table[0] for table in uvf.tablelist if table[1] == 'AIPS SN']
+                max_sn = max(sn_versions)
+                # Make sure we have a sequential integer sequence of SN table versions
+                assert sorted(sn_versions) == list(range(1, max_sn + 1))
+                pSN_get = min(max_sn, pSN)
+                if pSN_get == 0:
+                    raise ValueError('No self-calibration solutions available')
+                log.info('Extracting phase only self-calibration from AIPS SN: %d' % (pSN_get))
+                uvf.attach_table("AIPS SN", pSN_get)
+                sntab = uvf.tables["AIPS SN"]
```

Comment:
Because with these changes `pSN_get` will always point to the version number of an existing table in the UV data (Thus the `ValueError` on [L281](https://github.com/ska-sa/katsdpcontim/blob/063b6d28cc3494f8b2673946db0bf592023d1ff7/katacomb/katacomb/aips_export.py#L281) which essentially does what the `KeyError` check was doing before).

However I just noticed a problem. `sn_versions`  will be an empty list if there are no SN tables at all and then `max_sn` of [L276](https://github.com/ska-sa/katsdpcontim/blob/063b6d28cc3494f8b2673946db0bf592023d1ff7/katacomb/katacomb/aips_export.py#L276) will raise a cryptic `ValueError` instead. I'll add a `default=0` to ensure the `ValueError` on L281 is raised in this case (thanks Python 3).

Reply:
Good catch!